batch/batch/front_end/front_end.py

```diff
@@ -1489,6 +1499,89 @@ async def insert(tx):
+    state != 'complete' LIMIT 1) AS batch
+FROM billing_projects WHERE name = %s FOR UPDATE;
+    ''',
+            (billing_project,))
+        if not row:
+            raise KeyError(billing_project)
+        assert row['name'] == billing_project
+        if row['closed']:
+            raise ValueError(f'Billing project {billing_project} is already closed.')
+        if row['batch'] is not None:
+            raise RuntimeError(f'Billing project {billing_project} has open or running batches.')
+
+        await tx.execute_update(
+            'UPDATE billing_projects SET closed = TRUE WHERE name = %s;',
+            (billing_project, ))
+    await close_project()  # pylint: disable=no-value-for-parameter
+
+
+@routes.post('/billing_projects/{billing_project}/close')
+@prom_async_time(REQUEST_TIME_POST_CLOSE_BILLING_PROJECT_UI)
+@check_csrf_token
+@web_authenticated_developers_only(redirect=False)
+async def post_close_billing_projects(request, userdata):  # pylint: disable=unused-argument
+    db = request.app['db']
+    billing_project = request.match_info['billing_project']
+
+    session = await aiohttp_session.get_session(request)
+    try:
+        await _close_billing_project(db, billing_project)
+        set_message(session, f'Closed billing project {billing_project}.', 'info')
+    except KeyError as e:
+        set_message(session, f'Billing project {str(e)} does not exist.', 'error')
+    except RuntimeError as e:
+        set_message(session, str(e), 'error')
+    except ValueError as e:
+        set_message(session, str(e), 'info')
+    return web.HTTPFound(deploy_config.external_url('batch', '/billing_projects'))
+
+
+async def _reopen_billing_project(db, billing_project):
+    @transaction(db)
+    async def open_project(tx):
+        row = await tx.execute_and_fetchone(
+            'SELECT name, closed FROM billing_projects WHERE name = %s FOR UPDATE;',
+            (billing_project,))
+        if not row:
+            raise KeyError(billing_project)
+        assert row['name'] == billing_project
+        if not row['closed']:
+            raise ValueError(f'Billing project {billing_project} is already open.')
+
+        await tx.execute_update(
+            'UPDATE billing_projects SET closed = FALSE WHERE name = %s;',
+            (billing_project, ))
+    await open_project()  # pylint: disable=no-value-for-parameter
+
+
+@routes.post('/billing_projects/{billing_project}/reopen')
+@prom_async_time(REQUEST_TIME_POST_REOPEN_BILLING_PROJECT_UI)
+@check_csrf_token
+@web_authenticated_developers_only(redirect=False)
+async def post_reopen_billing_projects(request, userdata):  # pylint: disable=unused-argument
+    db = request.app['db']
+    billing_project = request.match_info['billing_project']
+
+    session = await aiohttp_session.get_session(request)
+    try:
+        await _reopen_billing_project(db, billing_project)
+        set_message(session, f'Re-opened billing project {billing_project}.', 'info')
```

Comment:
I think Dan wanted me in a different PR to put this not in the try except and put it in `else`

Reply:
Good call.